File: ios/SpellIt/DesignSystem.swift
import SwiftUI

// MARK: - Crayon palette (sRGB approximations of the site's oklch tokens)

extension Color {
    init(hex: UInt32) {
        self.init(
            .sRGB,
            red: Double((hex >> 16) & 0xFF) / 255,
            green: Double((hex >> 8) & 0xFF) / 255,
            blue: Double(hex & 0xFF) / 255,
        )
    }

    static let paper = Color(hex: 0xFDFBF2)
    static let ink = Color(hex: 0x31435C)
    static let mutedInk = Color(hex: 0x6E7B90)
    static let softBorder = Color(hex: 0xC3CCDA)
    static let secondaryBg = Color(hex: 0xF5F1E3)

    static let coral = Color(hex: 0xE2705F)
    static let coralSoft = Color(hex: 0xFADFD8)
    static let sun = Color(hex: 0xF0C452)
    static let sunSoft = Color(hex: 0xFAF0CF)
    static let leaf = Color(hex: 0x47B583)
    static let leafSoft = Color(hex: 0xDFF2E5)
    static let sky = Color(hex: 0x5A8FCB)
    static let skySoft = Color(hex: 0xDEEAF6)
    static let grape = Color(hex: 0x8659B5)
    static let grapeSoft = Color(hex: 0xEEE4F4)
}

// MARK: - Typography

extension Font {
    static func heading(_ size: CGFloat, weight: Font.Weight = .semibold) -> Font {
        .system(size: size, weight: weight, design: .rounded)
    }
}

// MARK: - Letter tile, the signature element

enum TileSize {
    case xs, sm, md, lg

    var side: CGFloat {
        switch self {
        case .xs: 32
        case .sm: 40
        case .md: 48
        case .lg: 56
        }
    }

    var fontSize: CGFloat {
        switch self {
        case .xs: 15
        case .sm: 19
        case .md: 24
        case .lg: 30
        }
    }

    static func forWord(_ word: String) -> TileSize {
        if word.count > 10 { return .xs }
        if word.count > 7 { return .sm }
        return .md
    }
}

/// Chunky letter tile with the hard offset shadow.
struct TileView: View {
    var letter: String
    var size: TileSize = .md
    var fill: Color = .white
    var dashed = false

    var body: some View {
        ZStack {
            if !dashed {
                RoundedRectangle(cornerRadius: 12)
                    .fill(Color.ink)
                    .offset(y: 4)
            }
            RoundedRectangle(cornerRadius: 12)
                .fill(dashed ? fill.opacity(0.5) : fill)
                .overlay(
                    RoundedRectangle(cornerRadius: 12)
                        .strokeBorder(
                            dashed ? Color.softBorder : Color.ink,
                            style: StrokeStyle(lineWidth: 3, dash: dashed ? [6, 5] : []),
                        )
                )
            Text(letter.uppercased())
                .font(.heading(size.fontSize, weight: .bold))
                .foregroundStyle(Color.ink)
                .minimumScaleFactor(0.5)
        }
        .frame(width: size.side, height: size.side)
        .accessibilityLabel(letter.isEmpty ? "blank" : letter)
    }
}

/// Tappable tile with a press-down animation.
struct TileButton: View {
    var letter: String
    var size: TileSize = .md
    var fill: Color = .white
    var disabled = false
    var action: () -> Void

    var body: some View {
        Button(action: action) {
            TileView(letter: letter, size: size, fill: fill)
                .opacity(disabled ? 0.35 : 1)
        }
        .buttonStyle(PressStyle())
        .disabled(disabled)
    }
}

struct PressStyle: ButtonStyle {
    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .offset(y: configuration.isPressed ? 3 : 0)
            .animation(.easeOut(duration: 0.12), value: configuration.isPressed)
    }
}

// MARK: - Chunky action buttons

struct ChunkyButtonStyle: ButtonStyle {
    var background: Color = .ink
    var foreground: Color = .white
    var bordered = false

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .font(.heading(16, weight: .semibold))
            .padding(.horizontal, 20)
            .padding(.vertical, 11)
            .background(
                ZStack {
                    RoundedRectangle(cornerRadius: 14)
                        .fill(Color.ink)
                        .offset(y: 3)
                    RoundedRectangle(cornerRadius: 14)
                        .fill(bordered ? Color.white : background)
                        .overlay(
                            RoundedRectangle(cornerRadius: 14)
                                .strokeBorder(Color.ink, lineWidth: bordered ? 2.5 : 0)
                        )
                }
            )
            .foregroundStyle(bordered ? Color.ink : foreground)
            .offset(y: configuration.isPressed ? 3 : 0)
            .animation(.easeOut(duration: 0.12), value: configuration.isPressed)
    }
}

// MARK: - Shake effect for wrong answers

struct ShakeEffect: GeometryEffect {
    var travel: CGFloat = 6
    var shakes: CGFloat = 3
    var animatableData: CGFloat

    func effectValue(size: CGSize) -> ProjectionTransform {
        ProjectionTransform(
            CGAffineTransform(
                translationX: travel * sin(animatableData * .pi * shakes * 2),
                y: 0,
            )
        )
    }
}

extension View {
    func shake(trigger: Int) -> some View {
        modifier(ShakeModifier(trigger: trigger))
    }
}

private struct ShakeModifier: ViewModifier {
    var trigger: Int

    func body(content: Content) -> some View {
        content
            .modifier(ShakeEffect(animatableData: CGFloat(trigger)))
            .animation(.easeInOut(duration: 0.35), value: trigger)
    }
}

File: android/app/src/main/java/com/skdaddle/spellit/ui/DesignSystem.kt
package com.skdaddle.spellit.ui

import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.material3.ripple
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.runtime.withFrameNanos
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.IntOffset
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlin.math.PI
import kotlin.math.sin

// Crayon palette (sRGB approximations of the site's oklch tokens)

object Palette {
    val Paper = Color(0xFFFDFBF2)
    val Ink = Color(0xFF31435C)
    val MutedInk = Color(0xFF6E7B90)
    val SoftBorder = Color(0xFFC3CCDA)
    val SecondaryBg = Color(0xFFF5F1E3)

    val Coral = Color(0xFFE2705F)
    val CoralSoft = Color(0xFFFADFD8)
    val Sun = Color(0xFFF0C452)
    val SunSoft = Color(0xFFFAF0CF)
    val Leaf = Color(0xFF47B583)
    val LeafSoft = Color(0xFFDFF2E5)
    val Sky = Color(0xFF5A8FCB)
    val SkySoft = Color(0xFFDEEAF6)
    val Grape = Color(0xFF8659B5)
    val GrapeSoft = Color(0xFFEEE4F4)
}

// Typography — one place to swap in a rounded display face later.

fun headingStyle(size: Int, weight: FontWeight = FontWeight.SemiBold): TextStyle =
    TextStyle(fontSize = size.sp, fontWeight = weight)

// Letter tile, the signature element

enum class TileSize(val side: Dp, val fontSize: Int) {
    XS(32.dp, 15),
    SM(40.dp, 19),
    MD(48.dp, 24),
    LG(56.dp, 30);

    companion object {
        fun forWord(word: String): TileSize = when {
            word.length > 10 -> XS
            word.length > 7 -> SM
            else -> MD
        }
    }
}

/** Chunky letter tile with the hard offset shadow. */
@Composable
fun Tile(
    letter: String,
    modifier: Modifier = Modifier,
    size: TileSize = TileSize.MD,
    fill: Color = Color.White,
    dashed: Boolean = false,
) {
    val shape = RoundedCornerShape(12.dp)
    Box(modifier = modifier.size(size.side)) {
        if (!dashed) {
            Box(
                Modifier
                    .matchParentSize()
                    .offset(y = 4.dp)
                    .background(Palette.Ink, shape),
            )
        }
        Box(
            Modifier
                .matchParentSize()
                .background(if (dashed) fill.copy(alpha = 0.5f) else fill, shape)
                .then(
                    if (dashed) {
                        Modifier.drawBehind {
                            drawRoundRect(
                                color = Palette.SoftBorder,
                                cornerRadius = CornerRadius(12.dp.toPx()),
                                style = Stroke(
                                    width = 3.dp.toPx(),
                                    pathEffect = PathEffect.dashPathEffect(
                                        floatArrayOf(6.dp.toPx(), 5.dp.toPx()),
                                    ),
                                ),
                            )
                        }
                    } else {
                        Modifier.border(3.dp, Palette.Ink, shape)
                    },
                ),
            contentAlignment = Alignment.Center,
        ) {
            Text(
                text = letter.uppercase(),
                style = headingStyle(size.fontSize, FontWeight.Bold),
                color = Palette.Ink,
                textAlign = TextAlign.Center,
                maxLines = 1,
            )
        }
    }
}

/** Tappable tile with a press-down animation. */
@Composable
fun TileButton(
    letter: String,
    modifier: Modifier = Modifier,
    size: TileSize = TileSize.MD,
    fill: Color = Color.White,
    enabled: Boolean = true,
    onClick: () -> Unit,
) {
    val interaction = remember { MutableInteractionSource() }
    val pressed by interaction.collectIsPressedAsState()
    val press by animateFloatAsState(
        targetValue = if (pressed) 3f else 0f,
        animationSpec = tween(durationMillis = 120),
        label = "tilePress",
    )
    Tile(
        letter = letter,
        size = size,
        fill = fill,
        modifier = modifier
            .offset { IntOffset(0, press.dp.roundToPx()) }
            .alpha(if (enabled) 1f else 0.35f)
            .clip(RoundedCornerShape(12.dp))
            .clickable(
                interactionSource = interaction,
                indication = null,
                enabled = enabled,
                onClick = onClick,
            ),
    )
}

// Chunky action button

@Composable
fun ChunkyButton(
    text: String,
    modifier: Modifier = Modifier,
    background: Color = Palette.Ink,
    foreground: Color = Color.White,
    bordered: Boolean = false,
    enabled: Boolean = true,
    onClick: () -> Unit,
) {
    val shape = RoundedCornerShape(14.dp)
    val interaction = remember { MutableInteractionSource() }
    val pressed by interaction.collectIsPressedAsState()
    val press by animateFloatAsState(
        targetValue = if (pressed) 3f else 0f,
        animationSpec = tween(durationMillis = 120),
        label = "buttonPress",
    )
    Box(
        modifier = modifier
            .alpha(if (enabled) 1f else 0.4f)
            .offset { IntOffset(0, press.dp.roundToPx()) },
    ) {
        Box(
            Modifier
                .matchParentSize()
                .offset(y = 3.dp)
                .background(Palette.Ink, shape),
        )
        Box(
            Modifier
                .clip(shape)
                .background(if (bordered) Color.White else background, shape)
                .then(if (bordered) Modifier.border(2.5.dp, Palette.Ink, shape) else Modifier)
                .clickable(
                    interactionSource = interaction,
                    indication = ripple(),
                    enabled = enabled,
                    onClick = onClick,
                )
                .padding(horizontal = 20.dp, vertical = 11.dp),
        ) {
            Text(
                text = text,
                style = headingStyle(16),
                color = if (bordered) Palette.Ink else foreground,
            )
        }
    }
}

// Shake effect for wrong answers

/**
 * Shakes its content horizontally whenever [trigger] changes to a new
 * non-zero value (mirrors the iOS ShakeEffect).
 */
@Composable
fun ShakeContainer(
    trigger: Int,
    modifier: Modifier = Modifier,
    content: @Composable () -> Unit,
) {
    val offsetX = remember { Animatable(0f) }
    LaunchedEffect(trigger) {
        if (trigger == 0) return@LaunchedEffect
        val durationMs = 350f
        val start = withFrameNanos { it }
        while (true) {
            val now = withFrameNanos { it }
            val t = ((now - start) / 1_000_000f) / durationMs
            if (t >= 1f) break
            offsetX.snapTo(6f * sin(t * PI.toFloat() * 3f * 2f))
        }
        offsetX.snapTo(0f)
    }
    Box(modifier.offset { IntOffset(offsetX.value.dp.roundToPx(), 0) }) {
        content()
    }
}
